### PR TITLE
Add port of AMD's Robust Contrast Adaptive Sharpening

### PR DESCRIPTION
**Pull request by Elabajaba**
# Objective

TAA, FXAA, and some other post processing effects can cause the image to become blurry. Sharpening helps to counteract that.

## Solution

~~This is a port of AMD's Contrast Adaptive Sharpening (I ported it from the [SweetFX](https://github.com/CeeJayDK/SweetFX/blob/master/Shaders/CAS.fx) version, which is still MIT licensed). CAS is a good sharpening algorithm that is better at avoiding the full screen oversharpening artifacts that simpler algorithms tend to create.~~

This is a port of AMD's Robust Contrast Adaptive Sharpening (RCAS) which they developed for FSR 1 ([and continue to use in FSR 2](https://github.com/GPUOpen-Effects/FidelityFX-FSR2/blob/149cf26e1229eaf5fecfb4428e71666cf4aee374/src/ffx-fsr2-api/shaders/ffx_fsr1.h#L599)). RCAS is a good sharpening algorithm that is better at avoiding the full screen oversharpening artifacts that simpler algorithms tend to create.

---

## Future Work

- Consider porting this to a compute shader for potentially better performance. (In my testing it is currently ridiculously cheap (0.01ms in Bistro at 1440p where I'm GPU bound), so this wasn't a priority, especially since it would increase complexity due to still needing the non-compute version for webgl2 support).

---

## Changelog

- Added Contrast Adaptive Sharpening.

### CODE REVIEW

File: crates/bevy_core_pipeline/src/contrast_adaptive_sharpening/mod.rs
Comment (JMS55) [inline crates/bevy_core_pipeline/src/contrast_adaptive_sharpening/mod.rs:27]: I'm not sure how people feel on a toggle to enable vs adding/removing the component. I'm fine with either one.

File: crates/bevy_core_pipeline/src/contrast_adaptive_sharpening/mod.rs
Comment (JMS55) [inline crates/bevy_core_pipeline/src/contrast_adaptive_sharpening/mod.rs:26]:
```suggestion
    /// Enable Sharpening.
```

File: crates/bevy_core_pipeline/src/contrast_adaptive_sharpening/mod.rs
Comment (JMS55) [inline crates/bevy_core_pipeline/src/contrast_adaptive_sharpening/mod.rs:21]: Needs a short paragraph describing the effect. Something about how it sharpens an image, which is useful in conjunction with effects like FXAA or TAA that soften things.

File: crates/bevy_core_pipeline/src/contrast_adaptive_sharpening/mod.rs
Comment (JMS55) [inline crates/bevy_core_pipeline/src/contrast_adaptive_sharpening/mod.rs:124]: This should maybe be tonemapping? In general, the way we order post processing probably needs improving.

**Review by JMS55 — CHANGES_REQUESTED** [review]
Awesome work, the added sharpness is a definite improvement :)

Would be nice to have this as a compute shader, with an IFDEF keeping the old version for WebGL. That can be a future PR though.

File: crates/bevy_core_pipeline/src/contrast_adaptive_sharpening/mod.rs
Comment (Elabajaba) [reply crates/bevy_core_pipeline/src/contrast_adaptive_sharpening/mod.rs:27]: I just copied what Fxaa was doing here. I don't really have a preference either way.

File: crates/bevy_core_pipeline/src/contrast_adaptive_sharpening/node.rs
Comment (Elabajaba) [reply crates/bevy_core_pipeline/src/contrast_adaptive_sharpening/node.rs:24]: I ended up shoving the uniforms into a Resource on extract (which, thinking about it, is wrong if we're going to let users have different settings per view) because I couldn't get it to accept the CASUniform struct when calling `create_bind_group` in `ContrastAdaptiveSharpeningNode::run()`.

File: crates/bevy_core_pipeline/src/contrast_adaptive_sharpening/mod.rs
Comment (Elabajaba) [reply crates/bevy_core_pipeline/src/contrast_adaptive_sharpening/mod.rs:124]: Tonemapping comes before FXAA, and sharpening should happen after FXAA, so that wouldn't work.

Comment (JMS55) [reply crates/bevy_core_pipeline/src/contrast_adaptive_sharpening/mod.rs:124]: Hmm, I apply TAA before tonemapping and bloom and some other stuff. Should CAS go immediately after TAA, or before tonemapping?

File: crates/bevy_core_pipeline/src/contrast_adaptive_sharpening/node.rs
Comment (JMS55) [reply crates/bevy_core_pipeline/src/contrast_adaptive_sharpening/node.rs:24]: Check out `UniformComponentPlugin`. The revamp bloom PR has an example of using it, and it's probably used elsewhere in bevy. Basically you can use that plugin, and then in the node you add a component to the view query, which gives you a dynamic offset that you set on the render pass. Feel free to ask me about it in more detail on discord.

File: crates/bevy_core_pipeline/src/contrast_adaptive_sharpening/mod.rs
Comment (Elabajaba) [reply crates/bevy_core_pipeline/src/contrast_adaptive_sharpening/mod.rs:131]: Sharpening should probably be one of the last post processing passes, to avoid any sharpening artifacts screwing up the output of a different pass. (Ideally it would happen after upscaling, if/when we get actual upscaling).

File: crates/bevy_core_pipeline/src/contrast_adaptive_sharpening/mod.rs
Comment (JMS55) [inline crates/bevy_core_pipeline/src/contrast_adaptive_sharpening/mod.rs:27]:
```suggestion
/// based on the local contrast. This can help avoid over-sharpening areas with high contrast
```

**Comment by Elabajaba** [comment]
What do people think about shortening the internal names from ContrastAdaptiveSharpening to CAS (eg. `ViewContrastAdaptiveSharpeningPipeline` -> `ViewCASPipeline`, `ContrastAdaptiveSharpeningNode` -> `CASNode`).

I think it's probably better to not change the user facing `ContrastAdaptiveSharpeningSettings` to `CASSettings` so that it keeps sharpening in the name.

**Comment by JMS55** [comment]
I like keeping the public API with the full name, but it's totally fine to shorten the internal type names. I did the same for TAA and SSAO.

We may even want to rename it to just `SharpeningSettings` or something - the contrast adaptive part could just be an implementation detail.

File: crates/bevy_core_pipeline/src/contrast_adaptive_sharpening/node.rs
Comment (Elabajaba) [reply crates/bevy_core_pipeline/src/contrast_adaptive_sharpening/node.rs]: I was already using `UniformComponentPlugin`, I just wasn't dynamically indexing into the resource containing it.

Fixing this also fixed a potential future bug where using it on multiple cameras with different values was broken.

**Review by JMS55 — APPROVED** [review]

**Comment by JMS55** [comment]
@Elabajaba needs rebasing due to stageless.

**Comment by Elabajaba** [comment]
Rebased, just not sure if I need to add `#[reflect(Component)]` to `ContrastAdaptiveSharpeningSettings` (like what is added in https://github.com/bevyengine/bevy/pull/7527).

Since this is also modifying the FXAA example (and TAA removes it), this either needs to be merged before TAA and then add the sharpening system into the new anti-aliasing example, or I can rebase it onto TAA and it can be merged after (or just change it to use the AA example once TAA is merged).

**Comment by alice-i-cecile** [comment]
@Elabajaba can you please resolve the comments you've addressed to help reviewers?

**Review by robtfm — CHANGES_REQUESTED** [review]
looks good, and shows a decent improvement with fxaa. i guess it'll be an even better impact with taa. i didn't grok the shader math but it looks in line with the reference.

it's not a matter for this PR, but i think we need to put some thought into node ordering. more nodes is making construction of core pipeline increasingly fragile; omitting a node (like FXAA) will cause ordering issues with nodes that order relative to it.
this isn't a problem as long as all these things are always in the core pipeline but it is not scalable and not open to non-core modification... something we can improve as part of rendergraph yeeting maybe.

**Comment by JMS55** [comment]
> it's not a matter for this PR, but i think we need to put some thought into node ordering. more nodes is making construction of core pipeline increasingly fragile; omitting a node (like FXAA) will cause ordering issues with nodes that order relative to it.
this isn't a problem as long as all these things are always in the core pipeline but it is not scalable and not open to non-core modification... something we can improve as part of rendergraph yeeting maybe.

Strongly agreed, that's part of what I opened the render node issues about. Imo, we should have labels be separate from nodes. So we always add labels and can order relative to them, but it doesn't matter if the node is in the graph or not. We should also pull out the node ordering into the core_3d file, instead of defining it per-plugin, which is more fragile.

File: crates/bevy_core_pipeline/src/contrast_adaptive_sharpening/node.rs
Comment (Elabajaba) [reply crates/bevy_core_pipeline/src/contrast_adaptive_sharpening/node.rs]: I rebased on main and pushed what should be a fix. I'll test it properly tomorrow.

**Comment by Elabajaba** [comment]
This has been updated to use Robust Contrast Adaptive Sharpening (RCAS) instead of CAS. RCAS is newer and was developed for FSR 1 (it's also used in FSR2), and uses fewer texture samples compared to CAS (5 instead of 9).

File: crates/bevy_core_pipeline/src/contrast_adaptive_sharpening/mod.rs
Comment (superdump) [inline crates/bevy_core_pipeline/src/contrast_adaptive_sharpening/mod.rs:51]: Perhaps it is useful if you have noisy textures?

File: crates/bevy_core_pipeline/src/contrast_adaptive_sharpening/robust_contrast_adaptive_sharpening.wgsl
Comment (superdump) [inline crates/bevy_core_pipeline/src/contrast_adaptive_sharpening/robust_contrast_adaptive_sharpening.wgsl:59]: Is this a requirement? Just asking because at some point I imagine we would add HDR display support and then perhaps the range should no longer be limited to 0,1, though I don't know.

File: crates/bevy_core_pipeline/src/contrast_adaptive_sharpening/robust_contrast_adaptive_sharpening.wgsl
Comment (superdump) [inline crates/bevy_core_pipeline/src/contrast_adaptive_sharpening/robust_contrast_adaptive_sharpening.wgsl:67]: Please use 4-space indentation throughout.

File: crates/bevy_core_pipeline/src/contrast_adaptive_sharpening/robust_contrast_adaptive_sharpening.wgsl
Comment (superdump) [inline crates/bevy_core_pipeline/src/contrast_adaptive_sharpening/robust_contrast_adaptive_sharpening.wgsl:87]: While mathematically equivalent, I wonder why AMD wrote it as `bB * 0.5 + (br * 0.5 + bG)`. Just wondering if there is some precision issue... OH, I know. It's to try to leverage fma (fused multiply-add). Your 'simplification' would be two additions and a multiplication. Their version would be two fmas. An fma is higher precision than other approaches.

```suggestion
   let bL = b.b * 0.5 + (b.r * 0.5 + b.g);
   let dL = d.b * 0.5 + (d.r * 0.5 + d.g);
   let eL = e.b * 0.5 + (e.r * 0.5 + e.g);
   let fL = f.b * 0.5 + (f.r * 0.5 + f.g);
   let hL = h.b * 0.5 + (h.r * 0.5 + h.g);
```

File: crates/bevy_core_pipeline/src/contrast_adaptive_sharpening/robust_contrast_adaptive_sharpening.wgsl
Comment (superdump) [inline crates/bevy_core_pipeline/src/contrast_adaptive_sharpening/robust_contrast_adaptive_sharpening.wgsl:38]: Please add a link to the reference code. This is what I found, but if you have a better link then please use that: https://github.com/GPUOpen-Effects/FidelityFX-FSR2/blob/ea97a113b0f9cadf519fbcff315cc539915a3acd/src/ffx-fsr2-api/shaders/ffx_fsr1.h#L672

File: crates/bevy_core_pipeline/src/contrast_adaptive_sharpening/robust_contrast_adaptive_sharpening.wgsl
Comment (superdump) [inline crates/bevy_core_pipeline/src/contrast_adaptive_sharpening/robust_contrast_adaptive_sharpening.wgsl:89]: fma:
```suggestion
    var noise = 0.25 * bL + 0.25 * dL + 0.25 * fL + 0.25 * hL - eL;
```

File: crates/bevy_core_pipeline/src/contrast_adaptive_sharpening/robust_contrast_adaptive_sharpening.wgsl
Comment (superdump) [inline crates/bevy_core_pipeline/src/contrast_adaptive_sharpening/robust_contrast_adaptive_sharpening.wgsl:91]: Hopefully the compiler is smart enough to realise this is an fma.

File: crates/bevy_core_pipeline/src/contrast_adaptive_sharpening/robust_contrast_adaptive_sharpening.wgsl
Comment (superdump) [inline crates/bevy_core_pipeline/src/contrast_adaptive_sharpening/robust_contrast_adaptive_sharpening.wgsl:95]: fma:
```suggestion
    return vec4<f32>((lobe * b + lobe * d + lobe * f + lobe * h + e.rgb) / (4.0 * lobe + 1.0), e.w);
```

**Review by superdump — CHANGES_REQUESTED** [review]
Minor changes, basically looks good. :) Thanks!

File: crates/bevy_core_pipeline/src/contrast_adaptive_sharpening/mod.rs
Comment (JMS55) [inline crates/bevy_core_pipeline/src/contrast_adaptive_sharpening/mod.rs:42]: This is kinda unintuitive - can we reverse it?

**Comment by JMS55** [comment]
Maybe we should add CAS to the TAA/FXAA bundles. We should at least link to CAS from the TAA docs, where it mentions the softness of TAA.

**Comment by superdump** [comment]
> Maybe we should add CAS to the TAA/FXAA bundles. We should at least link to CAS from the TAA docs, where it mentions the softness of TAA.

Sharpening by default makes sense if they are practically blurry. I didn't really feel that our TAA looked blurry with our implementation / configuration, but I haven't had chance to play with it much yet. And I think last time I checked, FXAA at higher quality settings didn't look blurry, but lower quality settings did. However... I have mostly been looking at it on a MacBook Pro with hidpi display, so things will naturally look sharper. I should test it without the winit scale factor 1.0 setting.

> Also, thoughts on calling the user-facing types just Sharpening?

I could imagine there being other sharpening methods long-term. I think it's better to keep specific implementations specifically-named and then when it becomes clear what the situation is, use an enum to group the variants. I don't think we're there yet though. So what I mean is, build the individual pieces of functionality with full flexibility, and then add convenience API on top when it makes sense to do so.

**Comment by Elabajaba** [comment]
Changed it to use a linear 0.0 to 1.0 value for sharpening strength, visually it looks pretty linear imo.

File: crates/bevy_core_pipeline/src/contrast_adaptive_sharpening/mod.rs
Comment (Elabajaba) [reply crates/bevy_core_pipeline/src/contrast_adaptive_sharpening/mod.rs]: Changed it to a linear 0.0 to 1.0.

File: crates/bevy_core_pipeline/src/contrast_adaptive_sharpening/robust_contrast_adaptive_sharpening.wgsl
Comment (Elabajaba) [reply crates/bevy_core_pipeline/src/contrast_adaptive_sharpening/robust_contrast_adaptive_sharpening.wgsl]: I don't know.

File: crates/bevy_core_pipeline/src/contrast_adaptive_sharpening/mod.rs
Comment (Elabajaba) [reply crates/bevy_core_pipeline/src/contrast_adaptive_sharpening/mod.rs:46]: The docs seemed to imply that it's mainly for when you're integrating it into an existing engine and you might not be able to insert it into the pipeline before eg. film grain is applied.

Comment (superdump) [reply crates/bevy_core_pipeline/src/contrast_adaptive_sharpening/mod.rs:46]: Ok. :)

**Review by superdump — APPROVED** [review]

**Review by JMS55 — APPROVED** [review]

**Review by robtfm — APPROVED** [review]

**Comment by DGriffin91** [comment]
@superdump 
> And I think last time I checked, FXAA at higher quality settings didn't look blurry, but lower quality settings did.

This is actually backwards. FXAA is both sharper and faster at lower settings. (Part of why I didn't use the word quality in the plugin, but sensitivity instead)

I agree that FXAA shouldn't enable CAS by default.

**Pull request merged**
